src/turtlebot3_drl/model/GPREDDY/sac_0_stage_1/_model_configuration_20260906-195259.txt
device = cuda
simulation_speed = 1
state_size = 44
action_size = 2
hidden_size = 512
input_size = 132
batch_size = 256
buffer_size = 1000000
discount_factor = 0.99
learning_rate = 0.001
tau = 0.003
step_time = 0.01
loss_function = <function smooth_l1_loss at 0x74552afa3a60>
epsilon = 1.0
epsilon_decay = 0.9995
epsilon_minimum = 0.05
reward_function = A
backward_enabled = False
stacking_enabled = True
stack_depth = 3
frame_skip = 4
networks = [SACActor(
  (fc1): Linear(in_features=132, out_features=512, bias=True)
  (fc2): Linear(in_features=512, out_features=512, bias=True)
  (mean): Linear(in_features=512, out_features=2, bias=True)
  (log_std): Linear(in_features=512, out_features=2, bias=True)
), SACCritic(
  (q1_state): Linear(in_features=132, out_features=256, bias=True)
  (q1_action): Linear(in_features=2, out_features=256, bias=True)
  (q1_fc): Linear(in_features=512, out_features=512, bias=True)
  (q1_out): Linear(in_features=512, out_features=1, bias=True)
  (q2_state): Linear(in_features=132, out_features=256, bias=True)
  (q2_action): Linear(in_features=2, out_features=256, bias=True)
  (q2_fc): Linear(in_features=512, out_features=512, bias=True)
  (q2_out): Linear(in_features=512, out_features=1, bias=True)
), SACCritic(
  (q1_state): Linear(in_features=132, out_features=256, bias=True)
  (q1_action): Linear(in_features=2, out_features=256, bias=True)
  (q1_fc): Linear(in_features=512, out_features=512, bias=True)
  (q1_out): Linear(in_features=512, out_features=1, bias=True)
  (q2_state): Linear(in_features=132, out_features=256, bias=True)
  (q2_action): Linear(in_features=2, out_features=256, bias=True)
  (q2_fc): Linear(in_features=512, out_features=512, bias=True)
  (q2_out): Linear(in_features=512, out_features=1, bias=True)
)]
iteration = 0
alpha = 0.2
target_entropy = -2.0
log_alpha = tensor([0.], device='cuda:0', requires_grad=True)
alpha_optimizer = Adam (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    eps: 1e-08
    lr: 0.001
    weight_decay: 0
)
actor_optimizer = AdamW (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    eps: 1e-08
    lr: 0.001
    weight_decay: 0.01
)
critic_optimizer = AdamW (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    eps: 1e-08
    lr: 0.001
    weight_decay: 0.01
)
last_actor_loss = 0.0
last_critic_loss = 0.0
last_alpha_loss = 0.0



src/turtlebot3_drl/model/GPREDDY/sac_0_stage_1/_model_configuration_20260908-135645.txt
device = cuda
simulation_speed = 1
state_size = 44
action_size = 2
hidden_size = 512
input_size = 132
batch_size = 256
buffer_size = 1000000
discount_factor = 0.99
learning_rate = 0.001
tau = 0.003
step_time = 0.01
loss_function = <function smooth_l1_loss at 0x73e2952d1a60>
epsilon = 1.0
epsilon_decay = 0.9995
epsilon_minimum = 0.05
reward_function = A
backward_enabled = False
stacking_enabled = True
stack_depth = 3
frame_skip = 4
networks = [SACActor(
  (fc1): Linear(in_features=132, out_features=512, bias=True)
  (fc2): Linear(in_features=512, out_features=512, bias=True)
  (mean): Linear(in_features=512, out_features=2, bias=True)
  (log_std): Linear(in_features=512, out_features=2, bias=True)
), SACCritic(
  (q1_state): Linear(in_features=132, out_features=256, bias=True)
  (q1_action): Linear(in_features=2, out_features=256, bias=True)
  (q1_fc): Linear(in_features=512, out_features=512, bias=True)
  (q1_out): Linear(in_features=512, out_features=1, bias=True)
  (q2_state): Linear(in_features=132, out_features=256, bias=True)
  (q2_action): Linear(in_features=2, out_features=256, bias=True)
  (q2_fc): Linear(in_features=512, out_features=512, bias=True)
  (q2_out): Linear(in_features=512, out_features=1, bias=True)
), SACCritic(
  (q1_state): Linear(in_features=132, out_features=256, bias=True)
  (q1_action): Linear(in_features=2, out_features=256, bias=True)
  (q1_fc): Linear(in_features=512, out_features=512, bias=True)
  (q1_out): Linear(in_features=512, out_features=1, bias=True)
  (q2_state): Linear(in_features=132, out_features=256, bias=True)
  (q2_action): Linear(in_features=2, out_features=256, bias=True)
  (q2_fc): Linear(in_features=512, out_features=512, bias=True)
  (q2_out): Linear(in_features=512, out_features=1, bias=True)
)]
iteration = 0
alpha = 0.2
target_entropy = -2.0
log_alpha = tensor([0.], device='cuda:0', requires_grad=True)
alpha_optimizer = Adam (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    eps: 1e-08
    lr: 0.001
    weight_decay: 0
)
actor_optimizer = AdamW (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    eps: 1e-08
    lr: 0.001
    weight_decay: 0.01
)
critic_optimizer = AdamW (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    eps: 1e-08
    lr: 0.001
    weight_decay: 0.01
)
last_actor_loss = 0.0
last_critic_loss = 0.0
last_alpha_loss = 0.0



src/turtlebot3_drl/turtlebot3_drl/common/settings.py
# ===================================================================== #
#                           GENERAL SETTINGS                            #
# ===================================================================== #

ENABLE_BACKWARD          = False    # Enable backward movement of the robot
ENABLE_STACKING          = True    # Enable processing multiple consecutive scan frames at every observation step
ENABLE_VISUAL            = False    # Meant to be used only during evaluation/testing phase
ENABLE_TRUE_RANDOM_GOALS = False    # If false, goals are selected semi-randomly from a list of known valid goal positions
ENABLE_DYNAMIC_GOALS     = False    # If true, goal difficulty (distance) is adapted according to current success rate
MODEL_STORE_INTERVAL     = 100      # Store the model weights every N episodes
GRAPH_DRAW_INTERVAL      = 10       # Draw the graph every N episodes (drawing too often will slow down training)
GRAPH_AVERAGE_REWARD     = 10       # Average the reward graph over every N episodes


# ===================================================================== #
#                         ENVIRONMENT SETTINGS                          #
# ===================================================================== #

# --- SIMULATION ENVIRONMENT SETTINGS ---
REWARD_FUNCTION = "B"           # Defined in reward.py
EPISODE_TIMEOUT_SECONDS = 50    # Number of seconds after which episode timeout occurs

TOPIC_SCAN = 'scan'
TOPIC_VELO = 'cmd_vel'
TOPIC_ODOM = 'odom'
ARENA_LENGTH                = 4.2   # meters
ARENA_WIDTH                 = 4.2   # meters
SPEED_LINEAR_MAX            = 0.22  # m/s
SPEED_ANGULAR_MAX           = 2.0   # rad/s

LIDAR_DISTANCE_CAP          = 3.5   # meters
THRESHOLD_COLLISION         = 0.13  # meters
THRESHOLD_GOAL              = 0.20  # meters

OBSTACLE_RADIUS             = 0.16  # meters
MAX_NUMBER_OBSTACLES        = 6
ENABLE_MOTOR_NOISE          = False # Add normally distributed noise to motor output to simulate hardware imperfections

# --- REAL ROBOT ENVIRONMENT SETTINGS ---
REAL_TOPIC_SCAN  = 'scan'
REAL_TOPIC_VELO  = 'cmd_vel'
REAL_TOPIC_ODOM  = 'odom'

# LiDAR density count your robot is providing
# NOTE: If you change this value you also have to modify
# NUM_SCAN_SAMPLES for the model in drl_environment.py
# e.g. if you increase this by 320 samples also increase
# NUM_SCAN_SAMPLES by 320 samples.
REAL_N_SCAN_SAMPLES         = 40

REAL_ARENA_LENGTH           = 4.2   # meters
REAL_ARENA_WIDTH            = 4.2   # meters
REAL_SPEED_LINEAR_MAX       = 0.22  # in m/s
REAL_SPEED_ANGULAR_MAX      = 2.0   # in rad/s

REAL_LIDAR_CORRECTION       = 0.40  # meters, subtracted from the real LiDAR values
REAL_LIDAR_DISTANCE_CAP     = 3.5   # meters, scan distances are capped this value
REAL_THRESHOLD_COLLISION    = 0.11  # meters, minimum distance to an object that counts as a collision
REAL_THRESHOLD_GOAL         = 0.35  # meters, minimum distance to goal that counts as reaching the goal


# ===================================================================== #
#                       DRL ALGORITHM SETTINGS                          #
# ===================================================================== #

# DRL parameters
ACTION_SIZE     = 2         # Not used for DQN, see DQN_ACTION_SIZE
HIDDEN_SIZE     = 512       # Number of neurons in hidden layers

BATCH_SIZE      = 256       # Number of samples per training batch
BUFFER_SIZE     = 1000000   # Number of samples stored in replay buffer before FIFO
DISCOUNT_FACTOR = 0.99
LEARNING_RATE   = 0.001
TAU             = 0.003

OBSERVE_STEPS   = 50000     # At training start random actions are taken for N steps for better exploration
STEP_TIME       = 0.01      # Delay between steps, can be set to 0
EPSILON_DECAY   = 0.9995    # Epsilon decay per step
EPSILON_MINIMUM = 0.05

# DQN parameters
DQN_ACTION_SIZE = 5
TARGET_UPDATE_FREQUENCY = 1000

# DDPG parameters

# TD3 parameters
POLICY_NOISE            = 0.2
POLICY_NOISE_CLIP       = 0.5
POLICY_UPDATE_FREQUENCY = 2

# Stacking
STACK_DEPTH = 3             # Number of subsequent frames processed per step
FRAME_SKIP  = 4             # Number of frames skipped in between subsequent frames

# Episode outcome enumeration
UNKNOWN = 0
SUCCESS = 1
COLLISION_WALL = 2
COLLISION_OBSTACLE = 3
TIMEOUT = 4
TUMBLE = 5
RESULTS_NUM = 6